src/main/java/com/mysticalchemy/crucible/CrucibleTile.java

```diff
@@ -173,34 +177,46 @@ public boolean tryAddIngredient(ItemStack stack) {
 		if (recipeManager == null) {
 			recipeManager = level.getRecipeManager();
 		}
-		if (recipeManager == null)
+		if (recipeManager == null) {
 			return false;
+		}
 
 		Optional<PotionIngredientRecipe> recipe = recipeManager
 				.getRecipesFor(RecipeInit.POTION_RECIPE_TYPE.get(), createDummyCraftingInventory(stack), level).stream()
 				.findFirst();
 
-		if (recipe.isPresent() && canMerge(recipe.get(), stack.getCount())) {
-			PotionIngredientRecipe resolved_recipe = recipe.get();
-			
-			//handle various properties
-			if (resolved_recipe.getMakesLingering())
-				is_lingering = true;
-			if (resolved_recipe.getMakesSplash())
-				is_splash = true;
-			if (resolved_recipe.getDurationAdded() > 0)
-				duration += resolved_recipe.getDurationAdded() * stack.getCount();
-			
-			heat = Mth.clamp(heat - ITEM_HEAT_LOSS * stack.getCount(), MIN_TEMP, MAX_TEMP);
-			
-			//effects
-			mergeEffects(resolved_recipe.getEffects(), stack.getCount());
-			recalculatePotionColor();
-			level.sendBlockUpdated(getBlockPos(), getBlockState(), getBlockState(), 3);
-			return true;
+		if (recipe.isEmpty()) {
+			return false;
 		}
 
-		return false;
+		PotionIngredientRecipe resolved_recipe = recipe.get();
+		CrucibleEvent.AddIngredient event = EventDispatcher.DispatchCrucibleAddIngredientEvent(this.getAllEffects(), resolved_recipe, stack);
+		if (event.getResult() == Event.Result.DENY) {
+			return false;
+		}
+
+		if (!canMerge(recipe.get(), stack.getCount()) && event.getResult() != Event.Result.ALLOW) {
+			return false;
+		}
+
+		//event may have modified the recipe
+		resolved_recipe = event.getRecipe();
```

Comment:
Is it possible after the event that the recipe could be empty?  Would a check be warranted here?

Reply:
Since the recipe is returned to being immutable, no this isn't possible.  This line may honestly be irrelevant if we're not going to allow recipe mutations.

I wonder if we shouldn't add a ModifiedRecipe that's only used internally, that we create on the fly rather than at parse time and which extends PotionIngredientRecipe.  That way the PotionIngredientRecipe can stay immutable, but a copy can be made of it to allow manipulation during the AddIngredient event.

Thoughts?